### DISCUSSION TITLE
Feature Releases for Version 2.x.x

### DISCUSSION THREAD

**Discussion by NickLiffen** [ScottBrenner/cfn-lint-action#44] [Ideas]
👋 I think it would be great to start to think about what features could be shipped in the `v2.x.x` release. I have thought of some and put them below, but I would love other peoples opinions and thoughts on what should/could be included. 

**Feature One - Self Hosted Runners Cache** 👏

For self-hosted runners, as they don't spin up a "new" machine every time, we are able to use the [tools-cache](https://github.com/actions/toolkit/tree/master/packages/tool-cache) action to cache the binary on the machine and then re-use it on the next run. This would stop the runner from downloading the "same" version of the tool on every run. This would speed up how long it takes to set up the `cfn-lint` tool. 

I don't know how many people who use this GitHub Action are using self-hosted runners, so it's a little hard to say what value this would provide. 

**Feature Two - GitHub.com Runners Cache** 👍

It would be great to come up with a way to cache the `cfn-lint` binary across workflow runs. We would have to do something funky as we would have to cache the `version number` of the binary and if that doesn't change then we just pull from the cache. The problem is going to be if the user **doesn't** specify a version number in their action, it uses `0.*`. We can't cache `0.*` as this would NEVER pull down the new version until a new major version was out. 

This one I think is going to be hard but is likely the best feature to add IMO. 

**Feature Three - Remove the `0.x` version pulling - create a helper that gets the latest version** 👻

Right now, if the user doesn't specify a version number, we simply pull down the latest `0.x` release. I think if ever we are going to get this working: **Feature Two - GitHub.com Runners Cache**, we are going to need to remove the `0.x` capability and create a helper function that returns the latest version of the `cfn-lint`  package and then uses that explicit version to download the package. 

This one is one of the easier features, but not a lot of upfront value, more behind the scenes work. 

**Feature Four - Opinionate Key Commands** 👾

I think it would be great if we were to opinionate some key commands into the `with` arguments of this GitHub Action. Right now we ask the user to specify the whole command they want to run. I think for key commands such as:

```
cfn-lint -t ./template.yml --ignore-checks W3011, E3030
```

We would create an opinionated command where the user uses this action like this:

```
      - name: Testing with CFN Lint Command
        uses: scottbrenner/cfn-lint-action@v2
        with:
          template: template.yml
          ignore: [W3011, E3030]
```

Now the foundation is there, I think this would be a good feature to have 👍 This is only one example, I think there are many that we could add. 

The value of this is the user doesn't have to specify whole commands, and it's all handled by the GitHub Action. However, more overhead on maintainers if these commands ever "break". 

**Feature Five - Integrations** 

Keen to know if people use this package (cfn-lint) with anything else, or maybe send the results of this scan to X service. I see a few other GitHub Actions doing similar things which allow the results of the scan to be sent to X tool/service for an aggregated experience. 

Unsure on the value of this one, just see other actions doing similar things to maybe worth us looking into 🤔

**Comment by liamraeAL** [comment]
+1 for `Feature Four - Opinionate Key Commands `

**Reply by liamraeAL** [reply]
Another feature that might be nice - once cfn-lint has run, it posts a report/summary in the PR as a comment.
e.g. `cfn-lint found 8 errors and 4 warnings`.
Saves a few seconds of navigating to the action, expanding the step and scrolling through results.

**Reply by NickLiffen** [reply]
What commands would you like to see? What's your use case?

**Reply by liamraeAL** [reply]
Just a report or summary of the results from the cfn-lint run. Our developers are still new to github actions, and the easier it is for them to view the lint results the better. 
The standard green tick from github is enough when things are all OK, but when things aren't OK its harder than it needs to be for developers to figure out why there's a big red X. 
Something as simple as 
`cfn-lint found 8 errors and 4 warnings in your file main.template` as a comment on the PR would suffice. Possibly using the `--format pretty` output option? 
It's just an suggestion 😄

**Reply by NickLiffen** [reply]
@liamraeAL I like the sound of it 👍 I will create a feature and let you know once it's in development 👍 I will scope it out a little bit and let you contribute 💯

**Comment by ScottBrenner** [comment]
I've just created https://github.com/ScottBrenner/cfn-lint-action/blob/main/.github/workflows/update-main-version.yml to ensure the ["v2" tag](https://github.com/ScottBrenner/cfn-lint-action/releases/tag/v2) is updated alongside every minor release!
